frame/undercollateralized-loans/src/lib.rs

```diff
@@ -0,0 +1,376 @@
+		TheLoanContractIsExpired,
+		// Tis should not happens.
+		// Error added for debug.
+		CollateralCanNotBeTransferedBackToTheBorrowersAccount,
+		// When we try to retrieve interest rate for the date which is not present in the payment
+		// schedule for particular loan.
+		ThereIsNoSuchMomentInTheLoanPaymentSchedule,
+	}
+
+	#[pallet::genesis_config]
+	#[derive(Default)]
+	pub struct GenesisConfig {}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			let now = T::UnixTime::now().as_secs() as Timestamp;
+			let current_date_timestamp = crate::Pallet::<T>::get_date_aligned_timestamp(now);
+			CurrentDateStorage::<T>::put(current_date_timestamp);
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl GenesisConfig {
+		/// Direct implementation of `GenesisBuild::build_storage`.
+		///
+		/// Kept in order not to break dependency.
+		pub fn build_storage<T: Config>(&self) -> Result<sp_runtime::Storage, String> {
+			<Self as frame_support::traits::GenesisBuild<T>>::build_storage(self)
+		}
+
+		/// Direct implementation of `GenesisBuild::assimilate_storage`.
+		///
+		/// Kept in order not to break dependency.
+		pub fn assimilate_storage<T: Config>(
+			&self,
+			storage: &mut sp_runtime::Storage,
+		) -> Result<(), String> {
+			<Self as frame_support::traits::GenesisBuild<T>>::assimilate_storage(self, storage)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		// TODO: @mikolaichuk: add weights calculation
+		fn on_initialize(block_number: T::BlockNumber) -> Weight {
+			Self::treat_vaults_balance(block_number);
+			let now = Self::now();
+			let stored_current_day_timestamp = CurrentDateStorage::<T>::get();
+			// Check if date is changed.
+			let stored_date = Self::date_from_timestamp(stored_current_day_timestamp);
+			let date = Self::date_from_timestamp(now);
+			if stored_date < date {
+				let current_date_aligned_timestamp = Self::get_date_aligned_timestamp(now);
+				CurrentDateStorage::<T>::put(current_date_aligned_timestamp);
+				Self::check_payments(current_date_aligned_timestamp);
+				// Terminate loans which were not activated by borrower before first payment date
+				// once a day.
+				Self::terminate_non_activated_expired_loans(current_date_aligned_timestamp);
+			}
+			1000
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(1000)]
+		#[transactional]
+		pub fn create_market(
+			origin: OriginFor<T>,
+			input: MarketInputOf<T>,
+			keep_alive: bool,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let market_info =
+				<Self as UndercollateralizedLoans>::create_market(who.clone(), input, keep_alive)?;
+			let event = Event::<T>::MarketCreated { market_info };
+			Self::deposit_event(event);
+			Ok(())
+		}
+
+		#[pallet::weight(1000)]
+		#[transactional]
+		pub fn create_loan(origin: OriginFor<T>, input: LoanInputOf<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(
```

Comment:
should this validity check be within the `UndercollateralizedLoans` implementation of the `create_loan` function

Reply:
@kevin-composable Lets assume that somebody wants to use trait's create_loan() implementation but without rights checking, in such case current approach would be more flexible.